crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,882 @@
+    }
+}
+
+/// Sort an `__all__` definition represented by a `StmtAnnAssign` AST node.
+/// For example: `__all__: list[str] = ["b", "c", "a"]`.
+pub(crate) fn sort_dunder_all_ann_assign(checker: &mut Checker, node: &ast::StmtAnnAssign) {
+    if let Some(value) = &node.value {
+        sort_dunder_all(checker, &node.target, value);
+    }
+}
+
+fn sort_dunder_all(checker: &mut Checker, target: &ast::Expr, node: &ast::Expr) {
+    let ast::Expr::Name(ast::ExprName { id, .. }) = target else {
+        return;
+    };
+
+    if id != "__all__" {
+        return;
+    }
+
+    // We're only interested in `__all__` in the global scope
+    if !checker.semantic().current_scope().kind.is_module() {
+        return;
+    }
+
+    let (elts, range, kind) = match node {
+        ast::Expr::List(ast::ExprList { elts, range, .. }) => (elts, *range, DunderAllKind::List),
+        ast::Expr::Tuple(tuple_node @ ast::ExprTuple { elts, range, .. }) => {
+            (elts, *range, DunderAllKind::Tuple(tuple_node))
+        }
+        _ => return,
+    };
+
+    let mut possibly_fixable = true;
+    let mut string_items = vec![];
+    for elt in elts {
+        // Don't flag `__all__` definitions that contain non-strings
+        let Some(string_literal) = elt.as_string_literal_expr() else {
+            return;
+        };
+        // If any strings are implicitly concatenated, don't bother trying to autofix
+        if string_literal.value.is_implicit_concatenated() {
+            possibly_fixable = false;
+        }
+        string_items.push(string_literal.value.to_str());
+    }
+    if dunder_all_is_already_sorted(&string_items) {
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(UnsortedDunderAll, range);
+
+    if possibly_fixable {
+        if let Some(fix) = get_fix(range, elts, &string_items, &kind, checker) {
+            diagnostic.set_fix(fix);
+        }
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+enum DunderAllKind<'a> {
+    List,
+    Tuple(&'a ast::ExprTuple),
+}
+
+impl DunderAllKind<'_> {
+    fn opening_token_for_multiline_definition(&self) -> Tok {
+        match self {
+            Self::List => Tok::Lsqb,
+            Self::Tuple(_) => Tok::Lpar,
+        }
+    }
+
+    fn closing_token_for_multiline_definition(&self) -> Tok {
+        match self {
+            Self::List => Tok::Rsqb,
+            Self::Tuple(_) => Tok::Rpar,
+        }
+    }
+}
+
+fn dunder_all_is_already_sorted(string_elements: &[&str]) -> bool {
+    let mut element_iter = string_elements.iter();
+    let Some(this) = element_iter.next() else {
+        return true;
+    };
+    let mut this_key = AllItemSortKey::from(*this);
+    for next in element_iter {
+        let next_key = AllItemSortKey::from(*next);
+        if next_key < this_key {
+            return false;
+        }
+        this_key = next_key;
+    }
+    true
+}
+
+struct AllItemSortKey {
+    category: InferredMemberType,
+    value: String,
+}
+
```

Comment:
We can avoid cloning here by using `self.items.sort_by` when sorting (I assume that's where you ran into lifetime issues)

```patch
Index: crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs b/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
--- a/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs	(revision d78e931ef1c7cf71946a4ce63ea5669aa2b27dac)
+++ b/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs	(date 1705398024594)
-struct AllItemSortKey {
+struct AllItemSortKey<'a> {
     category: InferredMemberType,
-    value: String,
+    value: &'a str,
 }
 
-impl Ord for AllItemSortKey {
+impl Ord for AllItemSortKey<'_> {
     fn cmp(&self, other: &Self) -> Ordering {
         self.category
             .cmp(&other.category)
@@ -233,31 +249,31 @@
     }
 }
 
-impl PartialOrd for AllItemSortKey {
+impl PartialOrd for AllItemSortKey<'_> {
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
         Some(self.cmp(other))
     }
 }
 
-impl PartialEq for AllItemSortKey {
+impl PartialEq for AllItemSortKey<'_> {
     fn eq(&self, other: &Self) -> bool {
         self.cmp(other) == Ordering::Equal
     }
 }
 
-impl Eq for AllItemSortKey {}
+impl Eq for AllItemSortKey<'_> {}
 
-impl From<&str> for AllItemSortKey {
-    fn from(value: &str) -> Self {
+impl<'a> From<&'a str> for AllItemSortKey<'a> {
+    fn from(value: &'a str) -> Self {
         Self {
             category: InferredMemberType::of(value),
-            value: String::from(value),
+            value,
         }
     }
 }
 
-impl From<&DunderAllItem> for AllItemSortKey {
-    fn from(item: &DunderAllItem) -> Self {
+impl<'a> From<&'a DunderAllItem> for AllItemSortKey<'a> {
+    fn from(item: &'a DunderAllItem) -> Self {
         Self::from(item.value.as_str())
     }
 }
@@ -434,7 +450,7 @@
         );
 
         self.items
-            .sort_by_cached_key(|item| AllItemSortKey::from(item));
+            .sort_by(|a, b| AllItemSortKey::from(a).cmp(&AllItemSortKey::from(b)));
         let joined_items = join_multiline_dunder_all_items(
             &self.items,
             locator,

```

Reply:
> We can avoid cloning here by using `self.items.sort_by` when sorting (I assume that's where you ran into lifetime issues)

That is indeed where I ran into lifetime issues, yup! TYVM